### ISSUE TITLE
valid-jsdoc with generated functions

### ISSUE THREAD

**Issue by vtfn** [eslint/eslint#7615]
I'm trying to use the `valid-jsdoc` rule to keep my doc comments consistent but most of my functions are evaluated from another function. Like this

```javascript
export default curry((arg1, arg2) => {
  // ....
}))
```
It seems that the parser does not recognize this code as a function, which is understandable as there is no way to it to know what the `curry` function returns.

JSDoc happens to have a `@function` tag that tells its compiler that the documented code is a function.
Would it be possible to make the eslint parser use that tag to make the following fail?

```javascript
/* eslint valid-jsdoc: ["error", { "prefer": { "arg": "param" } }] */

/**
 * @function
 * @arg {string} foo - Desc...
 */
export default curry((foo) => {
  //....
})
```

**Comment by kaicataldo** [comment]
It does look like our js-doc parser, Doctrine, supports the [@function tag](https://github.com/eslint/doctrine/blob/e33c6bb9b9b9cb33bed63c678347a33675776071/test/parse.js#L220), so this may be feasible without an upstream change.

**Comment by platinumazure** [comment]
Seems reasonable, I'm on board.

**Comment by nzakas** [comment]
@vtfn can you explain the functionality you're looking for? Most of what `valid-jsdoc` [does](http://eslint.org/docs/rules/valid-jsdoc#rule-details) wouldn't apply in this case. What is it that you're expecting this change to do?

**Comment by vtfn** [comment]
I'm using functions like `curry` to create the functions for a library I'm writing. These functions have their own spec and I am documenting them with jsdoc. As I said, the rule does not apply when building functions in this way but, although my functions are not created like the usual, like `function foo() { ... }`, they are functions, and I think it would be great if their docs could be linted as such.
I just want the parser to recognize doc comments with the tag `@function` as documentation for a function, so `valid-jsdoc` can validate them.
Sorry if I can't explain better than this.

**Comment by nzakas** [comment]
@vtfn Does that actually work with JSDoc? It [looks like](http://usejsdoc.org/tags-function.html) `@function` is meant to mark variables that are functions rather than what you're doing. What does JSDoc output for your example?

**Comment by vtfn** [comment]
Well. Right now, nothing. For now, I'm just writing the comments. 
But I think the `@function` tag is meant to work with more than just variables. Isn't the exported value of a module an object as well? Of course I might be wrong. If so, and if really there is no way to make the linter work with this sort of thing, thats finne, I just thought it would be greate to be able to lint the docs as docs for a function even if the object doesn't seem to be a one.

**Comment by nzakas** [comment]
Yeah, I'm afraid we just can't do anything here. Just to try to explain a bit more, the intended usage of `@function` is this:

```js
/** @function */
var paginate = paginateFactory(pages);
```

That would allow JSDoc to find `paginate` and mark it as a function. What it sounds like you're looking for is something that given the following code:

```js
/**
 * @function
 * @arg {string} foo - Desc...
 */
export default curry((foo) => {
  //....
})
```

you'd like to check the `@arg` against the arrow function, but there is no direct relationship between the JSDoc block and the arrow function. Because if you did this instead:

```js
/**
 * @function
 * @arg {string} foo - Desc...
 */
var someFunc = curry((foo) => {
  //....
})
```

then the JSDoc block applies to `someFunc`, rather than the arrow function.

So unfortunately, it doesn't look like this is something we can support. However, if you use this pattern frequently, you can always create a custom rule that does exactly what you want.
